### ISSUE TITLE
Instantiating cls is handled differently than instantiating the class

### ISSUE THREAD

**Issue by DavidLeber** [microsoft/pyright#5613]
**Describe the bug**
When attempting to instantiate a child of an abstract class that has not implemented all abstract methods, instantiating from `cls` does not raise an error. I would expect calling `cls` to behave exactly the same way as calling the class itself.

**Code or Screenshots**

```python
"""Demonstrate issue."""
from __future__ import annotations
from abc import ABC, abstractmethod

class DemoClass(ABC):
    """Abstract parent."""
    
    @abstractmethod
    @classmethod
    def generate(cls) -> DemoClass:
        """Generate a complex object."""
    
    @abstractmethod
    def needs_to_implement(self) -> None:
        """Child needs to implement."""

class FailingChild(DemoClass):
    """Child that fails to throw a Pyright error."""

    @classmethod
    def generate(cls) -> FailingChild:
        """Generate a complex object."""
        return cls()

class PassingChild(DemoClass):
    """Child that successfully throws a Pyright error."""
    
    @classmethod
    def generate(cls) -> PassingChild:
        """Generate a complex object."""
        return PassingChild()
```

**VS Code extension or command-line**
Running as a VSCode extension. Python version 3.10.8, Pyright version 1.1.316
